batch/batch/driver/pool.py

```diff
@@ -0,0 +1,469 @@
+                              worker_pd_ssd_data_disk_size_gb=self.worker_pd_ssd_data_disk_size_gb,
+                              worker_disk_size_gb=self.worker_disk_size_gb)
+
+    async def control_loop(self):
+        log.info(f'starting control loop for {self}')
+        while True:
+            try:
+                ready_cores = await self.db.select_and_fetchone(
+                    '''
+SELECT CAST(COALESCE(SUM(ready_cores_mcpu), 0) AS SIGNED) AS ready_cores_mcpu
+FROM user_inst_coll_resources
+WHERE inst_coll = %s
+LOCK IN SHARE MODE;
+''',
+                    (self.name,))
+
+                ready_cores_mcpu = ready_cores['ready_cores_mcpu']
+
+                free_cores_mcpu = sum([
+                    worker.free_cores_mcpu
+                    for worker in self.healthy_instances_by_free_cores
+                ])
+                free_cores = free_cores_mcpu / 1000
+
+                log.info(f'{self} n_instances {self.n_instances} {self.n_instances_by_state}'
+                         f' free_cores {free_cores} live_free_cores {self.live_free_cores_mcpu / 1000}'
+                         f' ready_cores {ready_cores_mcpu / 1000}')
+
+                if ready_cores_mcpu > 0 and free_cores < 500:
+                    n_live_instances = self.n_instances_by_state['pending'] + self.n_instances_by_state['active']
+                    instances_needed = (
+                        (ready_cores_mcpu - self.live_free_cores_mcpu + (self.worker_cores * 1000) - 1)
+                        // (self.worker_cores * 1000))
+                    instances_needed = min(instances_needed,
+                                           self.max_live_instances - n_live_instances,
+                                           self.max_instances - self.n_instances,
+                                           # 20 queries/s; our GCE long-run quota
+                                           300,
+                                           # n * 16 cores / 15s = excess_scheduling_rate/s = 10/s => n ~= 10
+                                           10)
+                    if instances_needed > 0:
+                        log.info(f'creating {instances_needed} new instances')
+                        # parallelism will be bounded by thread pool
+                        await asyncio.gather(*[self.create_instance() for _ in range(instances_needed)])
+
+                n_live_instances = self.n_instances_by_state['pending'] + self.n_instances_by_state['active']
+                if (ENABLE_STANDING_WORKER
+                        and self.standing_worker
+                        and n_live_instances == 0
+                        and self.max_instances > 0):
+                    await self.create_instance(cores=self.standing_worker_cores,
+                                               max_idle_time_msecs=STANDING_WORKER_MAX_IDLE_TIME_MSECS)
+            except asyncio.CancelledError:  # pylint: disable=try-except-raise
+                raise
```

Comment:
we do this pretty much everywhere we have manually written these loops. I think we should also do this in `retry_long_running`.